File: flash_newpipe_extractor/android/src/main/kotlin/kannel/outtis/flashEx/flash_newpipe_extractor/models/video.kt
package kannel.outtis.flashEx.flash_newpipe_extractor.models

import org.schabi.newpipe.extractor.stream.StreamInfoItem

class VideoInfo(private val item: StreamInfoItem){
    fun toMap():Map<String , Any?>{
        val itemMap: MutableMap<String, Any?> = mutableMapOf()
        itemMap["name"] = item.name
        itemMap["url"] = item.url
        itemMap["viewCount"] = item.viewCount
        itemMap["textualUploadDate"] = item.textualUploadDate
        itemMap["uploaderName"] = item.uploaderName
        itemMap["uploaderUrl"] = item.uploaderUrl
        itemMap["thumbnailUrl"] = item.thumbnailUrl
        itemMap["duration"] = item.duration
        itemMap["uploadDate"] = item.uploadDate!!.offsetDateTime().toString()
        itemMap["isUploaderVerified"] = item.isUploaderVerified
        return itemMap
    }
}

File: flash_newpipe_extractor/lib/src/models/video.dart
import 'package:flash_newpipe_extractor/src/method_calls.dart';
import 'package:flash_newpipe_extractor/src/models/videoInfo.dart';

import '../utils/utils.dart';
import 'channel.dart';

class YoutubeVideo {
  final String? videoName;
  final String url;
  final int? viewCount;
  final String? textualUploadDate;
  final String? uploaderName;
  final String? uploaderUrl;
  final String? thumbnailUrl;
  final Duration? duration;
  final DateTime? uploadDate;
  final bool? isUploaderVerified;

  YoutubeVideo({
    this.videoName,
    required this.url,
    this.viewCount,
    this.textualUploadDate,
    this.uploaderName,
    this.uploaderUrl,
    this.thumbnailUrl,
    this.duration,
    this.uploadDate,
    this.isUploaderVerified,
  });
  String get id => Utils.getIdFromUrl(url);

  YoutubeVideoInfo? _videoInfo;
  YoutubeVideoInfo? get videoInfo => _videoInfo;

  Future<YoutubeVideoInfo> get getFullInformation async {
    return _videoInfo = await FlashMethodCalls.getVideoInfoFromUrl(url);
  }

  String get hqdefault => "https://img.youtube.com/vi/$id/hqdefault.jpg";

  String get mqdefault => "https://img.youtube.com/vi/$id/mqdefault.jpg";

  String get sddefault => "https://img.youtube.com/vi/$id/sddefault.jpg";

  String get maxresdefault =>
      "https://img.youtube.com/vi/$id/maxresdefault.jpg";

  Channel? _uploaderChannel;
  Channel? get uploaderChannelInfo => _uploaderChannel;
  Future<Channel?> getUploaderChannelInfo() async {
    try {
      return _uploaderChannel =
          await FlashMethodCalls.getChannelInfo(uploaderUrl!);
    } catch (e) {
      return null;
    }
  }

  factory YoutubeVideo.fromMap(Map<String, dynamic> map) {
    return YoutubeVideo(
      url: map["url"],
      uploaderUrl: map["uploaderUrl"],
      isUploaderVerified: map["isUploaderVerified"],
      textualUploadDate: map["textualUploadDate"],
      uploadDate: DateTime.tryParse(
        Utils.retrieveTime(
          map["uploadDate"],
        ),
      ),
      duration: Duration(seconds: map["duration"]),
      thumbnailUrl: map["thumbnailUrl"],
      uploaderName: map["uploaderName"],
      videoName: map["name"],
      viewCount: map["viewCount"],
    );
  }
}
